File: web/src/lib/scoring.ts
/**
 * Standings -- computed, never stored (LLD §7).
 *
 *   score       = sum of question.score over questions the owner has a current AC on
 *   solve_time  = earliest AC submission time - awarded_at   ("first AC only counts")
 *   rank        = score DESC, total_time ASC, phase1_rank ASC
 *
 * A rejudge changes verdicts, and standings follow with nothing to recompute.
 * For 20 participants this is a few small queries and a sort.
 */
import { and, eq, isNull, lt, sql } from "drizzle-orm";

import { db } from "@/db";
import { user } from "@/db/auth-schema";
import { judgement, ownership, participant, question, submission } from "@/db/schema";

import { getContest } from "./contest";
import { phase1Standings } from "./phase1-review";

export type Standing = {
  participant_id: string;
  name: string;
  score: number;
  solved: number;
  total_time_ms: number;
  phase1_rank: number | null;
  rank: number;
};

export async function phase2Standings(opts: { frozenAt?: Date | null } = {}): Promise<Standing[]> {
  const people = await db
    .select({ id: participant.userId, name: user.name, disqualifiedAt: participant.disqualifiedAt })
    .from(participant)
    .innerJoin(user, eq(user.id, participant.userId));

  // Every current AC, with its submission time and the question's score and award time.
  const acs = await db
    .select({
      participantId: submission.participantId,
      questionId: submission.questionId,
      createdAt: submission.createdAt,
      score: question.score,
      awardedAt: ownership.awardedAt,
    })
    .from(judgement)
    .innerJoin(submission, eq(submission.id, judgement.submissionId))
    .innerJoin(question, eq(question.id, submission.questionId))
    .innerJoin(ownership, and(eq(ownership.questionId, submission.questionId), eq(ownership.participantId, submission.participantId)))
    .where(
      and(
        eq(judgement.verdict, "AC"),
        isNull(judgement.supersededAt),
        isNull(ownership.voidedAt),
        sql`${question.status} <> 'void'`,
        opts.frozenAt ? lt(submission.createdAt, opts.frozenAt) : undefined,
      ),
    );

  // First AC per (participant, question): MIN over submission time.
  const first = new Map<string, { score: number; solveMs: number }>();
  for (const a of acs) {
    const key = `${a.participantId}|${a.questionId}`;
    const solveMs = a.createdAt.getTime() - a.awardedAt.getTime();
    const prev = first.get(key);
    if (!prev || solveMs < prev.solveMs) first.set(key, { score: a.score, solveMs });
  }

  const p1 = await phase1Standings();
  const p1Rank = new Map(p1.map((s) => [s.participant_id, s.rank]));

  const rows: Standing[] = people
    .filter((p) => !p.disqualifiedAt)
    .map((p) => {
      let score = 0, solved = 0, total = 0;
      for (const [key, v] of first) {
        if (!key.startsWith(`${p.id}|`)) continue;
        score += v.score; solved += 1; total += v.solveMs;
      }
      return { participant_id: p.id, name: p.name, score, solved, total_time_ms: total, phase1_rank: p1Rank.get(p.id) ?? null, rank: 0 };
    });

  rows.sort((a, b) =>
    b.score - a.score ||
    a.total_time_ms - b.total_time_ms ||
    (a.phase1_rank ?? Infinity) - (b.phase1_rank ?? Infinity) ||
    a.name.localeCompare(b.name));

  // Shared rank only when everything is equal, including Phase 1 rank (US-B7-02).
  let rank = 0;
  rows.forEach((r, i) => {
    const prev = rows[i - 1];
    const tied = prev && prev.score === r.score && prev.total_time_ms === r.total_time_ms && prev.phase1_rank === r.phase1_rank;
    r.rank = tied ? prev.rank : (rank = i + 1);
  });
  return rows;
}

/** The leaderboard as participants may see it, honouring the administrator's setting. */
export async function leaderboardForParticipants() {
  const c = await getContest();
  if (c.leaderboardMode === "hidden") return { mode: "hidden" as const, standings: [] as Standing[], frozen_at: null };
  const frozenAt = c.leaderboardMode === "frozen" ? c.leaderboardFrozenAt : null;
  return { mode: c.leaderboardMode, standings: await phase2Standings({ frozenAt }), frozen_at: frozenAt?.toISOString() ?? null };
}

File: engine/engine/coding/scoring.py
"""Phase 2 standings -- computed on every read, never stored.

    score      = sum of question.score over questions the owner has a current AC on
    solve_time = earliest AC submission time - awarded_at   ("first AC only counts")
    rank       = score DESC, total time ASC, Phase 1 rank ASC

Because nothing is stored, a rejudge, a void or a take-back changes the
standings with nothing to recompute and no row that can go stale. There is no
leaderboard state to race over: concurrent readers just compute the same answer.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

import sqlalchemy as sa

from engine.contest.rules import get_contest
from engine.core import clock, db
from engine.phase1.standings import phase1_standings
from engine.schema import judgement, ownership, participant, question, submission, user


def phase2_standings(
    conn: sa.Connection, frozen_at: datetime | None = None
) -> list[dict[str, Any]]:
    people = conn.execute(
        sa.select(participant.c.user_id, user.c.name)
        .join(user, user.c.id == participant.c.user_id)
        .where(participant.c.disqualified_at.is_(None))
    ).all()
    first_solves = _first_solves(conn, frozen_at)
    p1_rank = {s["participant_id"]: s["rank"] for s in phase1_standings(conn)}
    rows: list[dict[str, Any]] = []
    for p in people:
        solves = first_solves.get(p.user_id, [])
        rows.append(_standing(p, solves, p1_rank.get(p.user_id)))
    rows.sort(
        key=lambda r: (
            -r["score"],
            r["total_time_ms"],
            r["phase1_rank"] if r["phase1_rank"] is not None else float("inf"),
            r["name"],
        )
    )
    _assign_ranks(rows)
    return rows


def _first_solves(
    conn: sa.Connection, frozen_at: datetime | None
) -> dict[str, list[tuple[int, int]]]:
    """Per participant, one (score, solve_ms) per question: their earliest current AC on it."""
    query = (
        sa.select(
            submission.c.participant_id,
            submission.c.question_id,
            submission.c.created_at,
            question.c.score,
            ownership.c.awarded_at,
        )
        .join(submission, submission.c.id == judgement.c.submission_id)
        .join(question, question.c.id == submission.c.question_id)
        .join(
            ownership,
            sa.and_(
                ownership.c.question_id == submission.c.question_id,
                ownership.c.participant_id == submission.c.participant_id,
            ),
        )
        .where(
            judgement.c.verdict == "AC",
            judgement.c.superseded_at.is_(None),
            ownership.c.voided_at.is_(None),
            question.c.status != "void",
        )
    )
    if frozen_at is not None:
        query = query.where(submission.c.created_at < frozen_at)
    best: dict[tuple[str, str], tuple[int, int]] = {}
    for ac in conn.execute(query):
        solve_ms = clock.ms(ac.created_at) - clock.ms(ac.awarded_at)
        key = (ac.participant_id, ac.question_id)
        if key not in best or solve_ms < best[key][1]:
            best[key] = (ac.score, solve_ms)
    out: dict[str, list[tuple[int, int]]] = {}
    for (pid, _), solve in best.items():
        out.setdefault(pid, []).append(solve)
    return out


def _standing(p: sa.Row, solves: list[tuple[int, int]], p1_rank: int | None) -> dict[str, Any]:
    return {
        "participant_id": p.user_id,
        "name": p.name,
        "score": sum(score for score, _ in solves),
        "solved": len(solves),
        "total_time_ms": sum(ms for _, ms in solves),
        "phase1_rank": p1_rank,
        "rank": 0,
    }


def _assign_ranks(rows: list[dict[str, Any]]) -> None:
    """A shared rank only when everything that breaks ties is equal."""
    tie_keys = ("score", "total_time_ms", "phase1_rank")
    for i, r in enumerate(rows):
        prev = rows[i - 1] if i else None
        if prev and all(prev[k] == r[k] for k in tie_keys):
            r["rank"] = prev["rank"]
        else:
            r["rank"] = i + 1


def frozen_at(c: sa.Row) -> datetime | None:
    return c.leaderboard_frozen_at if c.leaderboard_mode == "frozen" else None


def for_participants() -> dict[str, Any]:
    """The leaderboard as participants may see it, honouring the organisers' setting."""
    with db.transaction() as conn:
        c = get_contest(conn)
        if c.leaderboard_mode == "hidden":
            return {"mode": "hidden", "standings": [], "frozen_at": None}
        return {
            "mode": c.leaderboard_mode,
            "standings": phase2_standings(conn, frozen_at(c)),
            "frozen_at": clock.iso(frozen_at(c)),
        }


def for_staff() -> dict[str, Any]:
    """Both boards in full, ignoring the hidden/frozen setting."""
    with db.transaction() as conn:
        return {
            "phase1": phase1_standings(conn, with_contact=True),
            "phase2": phase2_standings(conn),
        }
